jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java

```diff
@@ -750,24 +750,28 @@ static void checkClassModuleVisibility(int accessMode, Module accessModule, Clas
 
 			if (INTERNAL_PRIVILEGED != accessMode) {
 				Module targetModule = targetClass.getModule();
-				String targetClassPackageName = targetClass.getPackageName();
-				if ((UNCONDITIONAL & accessMode) == UNCONDITIONAL) {
-					/* publicLookup objects can see all unconditionally exported packages. */
-					if (!targetModule.isExported(targetClassPackageName)) {
-						/*[MSG "K0676", "Module '{0}' no access to: package '{1}' which is not exported by module '{2}'"]*/
-						throw new IllegalAccessException(Msg.getString("K0587", getModuleName(accessModule), targetClassPackageName, getModuleName(targetModule))); //$NON-NLS-1$
-					}
-				} else if (!(
-					Objects.equals(accessModule, targetModule)
-					|| (targetModule.isExported(targetClassPackageName, accessModule) && accessModule.canRead(targetModule)))
-				) {
-					if (!targetModule.isExported(targetClassPackageName, accessModule)) {
-						/*[MSG "K0677", "Module '{0}' no access to: package '{1}' which is not exported by module '{2}' to module '{0}'"]*/
-						throw new IllegalAccessException(Msg.getString("K0677", getModuleName(accessModule), targetClassPackageName, getModuleName(targetModule))); //$NON-NLS-1$
+				// modules may be null during bootstrapping
+				if ((null != targetModule) && (null != accessModule)) { 
+					String targetClassPackageName = targetClass.getPackageName();
+					if ((UNCONDITIONAL & accessMode) == UNCONDITIONAL) {
+						/* publicLookup objects can see all unconditionally exported packages. */
+						if (!targetModule.isExported(targetClassPackageName)) {
+							/*[MSG "K0676", "Module '{0}' no access to: package '{1}' which is not exported by module '{2}'"]*/
+							throw new IllegalAccessException(Msg.getString("K0587", getModuleName(accessModule), targetClassPackageName, getModuleName(targetModule))); //$NON-NLS-1$
```

Comment:
There are three `throw new Illegal...` clauses in this method which generates a lot of bytecode.  It's often cleaner to pull this out into a method like:

```java
IllegalAccessException throwNewIAEForModuleAccess(String message, Module accessModule, String targetClassPackageName, Module targetModule) {
 throw new IllegalAccessException(Msg.getString(message, getModuleName(accessModule), targetClassPackageName, getModuleName(targetModule)));
}
```
and then call with code like:
```java
throw throwNewIAEForModuleAccess(.....);
```
as this makes it clear to the reader that execution can't continue past that call.

Reply:
Also legacy code, but I really like your idea.